common/src/gas_provider/negative_imbalance.rs

```diff
@@ -17,111 +17,62 @@
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
 use super::*;
-use core::cmp::Ordering;
 
 /// Opaque, move-only struct with private field to denote that value has been
 /// destroyed without any equal and opposite accounting.
 #[derive(RuntimeDebug, PartialEq, Eq)]
-pub struct NegativeImbalance<Balance: BalanceTrait, TotalValue: ValueStorage<Value = Balance>>(
-    Balance,
-    PhantomData<TotalValue>,
-);
-
-impl<Balance: BalanceTrait, TotalValue: ValueStorage<Value = Balance>>
-    NegativeImbalance<Balance, TotalValue>
-{
+pub struct NegativeImbalance<Balance: BalanceTrait>(Balance);
+
+impl<Balance: BalanceTrait> NegativeImbalance<Balance> {
     /// Create a new negative imbalance from value amount.
     pub fn new(amount: Balance) -> Self {
-        NegativeImbalance(amount, PhantomData)
-    }
-}
-
-impl<Balance: BalanceTrait, TotalValue: ValueStorage<Value = Balance>> TryDrop
-    for NegativeImbalance<Balance, TotalValue>
-{
-    fn try_drop(self) -> Result<(), Self> {
-        self.drop_zero()
-    }
-}
-
-impl<Balance: BalanceTrait, TotalValue: ValueStorage<Value = Balance>> Default
-    for NegativeImbalance<Balance, TotalValue>
-{
-    fn default() -> Self {
-        Self::zero()
+        NegativeImbalance(amount)
     }
 }
 
-impl<Balance: BalanceTrait, TotalValue: ValueStorage<Value = Balance>> Imbalance<Balance>
-    for NegativeImbalance<Balance, TotalValue>
-{
-    type Opposite = PositiveImbalance<Balance, TotalValue>;
+impl<Balance: BalanceTrait> Imbalance for NegativeImbalance<Balance> {
+    type Balance = Balance;
 
-    fn zero() -> Self {
-        Self(Zero::zero(), PhantomData)
+    fn peek(&self) -> Self::Balance {
+        self.0
     }
 
-    fn drop_zero(self) -> Result<(), Self> {
-        if self.0.is_zero() {
-            Ok(())
-        } else {
-            Err(self)
+    fn apply_to(&self, amount: &mut Option<Balance>) {
+        let new_value = amount.unwrap_or_else(Zero::zero);
+        if self.0 > new_value {
+            log::debug!(
```

Comment:
please create issue to make this fallible, cause such branches are critical errors for the system